internal/command/postgres/failover.go

```diff
@@ -128,3 +141,230 @@ func runFailover(ctx context.Context) (err error) {
+		}
+	}
+	if primary_region == "" {
+		return fmt.Errorf("Could not find primary region for app")
+	}
+
+	newLeader, err := pickNewLeader(ctx, app, machines_within_primary_region)
+	if err != nil {
+		return err
+	}
+
+	// Stop the leader
+	fmt.Println("Stopping current leader... ", leader.ID)
+	machineStopInput := api.StopMachineInput{
+		ID:     leader.ID,
+		Signal: "SIGINT",
+	}
+
+	flapsClient := flaps.FromContext(ctx)
+	err = flapsClient.Stop(ctx, machineStopInput, leader.LeaseNonce)
+	if err != nil {
+		return fmt.Errorf("could not stop pg leader %s: %w", leader.ID, err)
+	}
+
+	fmt.Println("Starting new leader")
+	_, err = flapsClient.Start(ctx, newLeader.ID, newLeader.LeaseNonce)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Promoting new leader... ", newLeader.ID)
+	err = ssh.SSHConnect(&ssh.SSHParams{
+		Ctx:      ctx,
+		Org:      app.Organization,
+		App:      app.Name,
+		Username: "postgres",
+		Dialer:   agent.DialerFromContext(ctx),
+		Cmd:      "repmgr standby promote --siblings-follow -f /data/repmgr.conf",
+		Stdout:   ioutils.NewWriteCloserWrapper(colorable.NewColorableStdout(), func() error { return nil }),
+		Stderr:   ioutils.NewWriteCloserWrapper(colorable.NewColorableStderr(), func() error { return nil }),
+		Stdin:    nil,
+	}, newLeader.PrivateIP)
+	if err != nil {
+		return err
+
+	}
+
+	// Restart the old leader
+	fmt.Fprintf(io.Out, "Restarting old leader... %s\n", leader.ID)
+	mach, err := flapsClient.Start(ctx, leader.ID, leader.LeaseNonce)
+	if err != nil {
+		return err
+	}
+	if mach.Status == "error" {
+		return fmt.Errorf("old leader %s could not be started: %s", leader.ID, mach.Message)
+	}
+
+	fmt.Printf("Waiting for leadership to swap to %s...\n", newLeader.ID)
+	if err := retry.Do(
+		func() error {
+			leader, err = flapsClient.Get(ctx, newLeader.ID)
+			if err != nil {
+				return err
+			}
+
+			if isLeader(leader) {
+				return nil
+			} else {
+				return fmt.Errorf("Machine %s never became the leader", newLeader.ID)
+			}
+
+		},
+		retry.Context(ctx), retry.Attempts(60), retry.Delay(time.Second), retry.DelayType(retry.FixedDelay),
+	); err != nil {
+		return err
+	}
+
+	// wait for health checks to pass
+	if err := watch.MachinesChecks(ctx, machines); err != nil {
+		return fmt.Errorf("failed to wait for health checks to pass: %w", err)
+	}
+
+	fmt.Fprintf(io.Out, "Failover complete\n")
+	return nil
+}
+
+func handleFlexFailoverFail(ctx context.Context, machines []*api.Machine) (err error) {
+	io := iostreams.FromContext(ctx)
+	flapsClient := flaps.FromContext(ctx)
+
+	leader, err := pickLeader(ctx, machines)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(io.ErrOut, "Error promoting new leader, restarting existing leader")
+	fmt.Println("Waiting for old leader to finish stopping")
+	if err := retry.Do(
+		func() error {
+			leader, err = flapsClient.Get(ctx, leader.ID)
+			if err != nil {
+				return err
+			}
+
+			if leader.State == "stopped" || leader.State == "started" {
```

Comment:
This bit doesn't quite register with me.  If we are waiting for the leader to stop, why are we returning nil if it's still in a started state?

Reply:
I should explain this in a comment, that's entirely my bad. So basically, because of the fact that we have to handle a failover fail at any time, it's possible that the leader hasn't even been stopped yet before failure (due to `pickNewLeader` failing). If that happens, there's no reason to try and stop that machine again, just to start it. The only reason we wait for it to finish stopping at all is we'll get an error if we don't.